system/lib/pthread/pthread_create.c

```diff
@@ -48,9 +55,36 @@ int __pthread_create(pthread_t *thread, const pthread_attr_t *attr, void *(*star
   return __pthread_create_js(thread, attr, start_routine, arg);
 }
 
-void _emscripten_thread_exit(void* retval) {
-  assert(pthread_self());
-  __pthread_exit_js(retval);
+void _emscripten_thread_exit(void* result) {
+  struct pthread *self = __pthread_self();
+  assert(self);
+
+  self->canceldisable = PTHREAD_CANCEL_DISABLE;
+  self->cancelasync = PTHREAD_CANCEL_DEFERRED;
+  self->result = result;
+
+  __pthread_exit_run_handlers();
+
+  // Call into the musl function that runs destructors of all thread-specific data.
+  __pthread_tsd_run_dtors();
+
+  if (self == emscripten_main_browser_thread_id()) {
+    // FIXME(sbc): When pthread_exit causes the entire application to exit
+    // we should be returning zero (according to the man page for pthread_exit).
+    exit((intptr_t)result);
+    return;
+  }
+
+  // Mark the thread as no longer running.
+  // When we publish this, the main thread is free to deallocate the thread object and we are done.
+  // Therefore set _pthread_self = 0; above to 'release' the object in this worker thread.
```

Comment:
Where is the thing this comment mentions done?

Reply:
I think this is referring to the _emscripten_thread_init below.  The comment is just copied directly and equally inaccurate before this change.

Reply:
I just removed this line since _emscripten_thread_init is already commented below.